sound/soc/intel/boards/sof_sdw.c

```diff
@@ -409,27 +409,23 @@ static int create_codec_dai_name(struct device *dev,
 	return 0;
 }
 
-static int set_codec_init_func(const struct snd_soc_acpi_link_adr *link,
+static int set_dailink_init_func(const struct snd_soc_acpi_link_adr *link,
 			       struct snd_soc_dai_link *dai_links,
 			       bool playback)
 {
-	int i;
-
-	for (i = 0; i < link->num_adr; i++) {
-		unsigned int part_id;
-		int codec_index;
+	unsigned int part_id;
+	int codec_index;
 
-		part_id = SDW_PART_ID(link->adr_d[i].adr);
-		codec_index = find_codec_info_part(part_id);
+	part_id = SDW_PART_ID(link->adr_d[0].adr);
+	codec_index = find_codec_info_part(part_id);
 
-		if (codec_index < 0)
-			return codec_index;
+	if (codec_index < 0)
+		return codec_index;
 
-		if (codec_info_list[codec_index].init)
-			codec_info_list[codec_index].init(link, dai_links,
```

Comment:
I don't understand the process well enough, but this raises an alarm for me. Are we really sure that it's always enough to just call `.init()` once for the first codec iinfo? That we'll never have mixed codecs or a codec, that needs certain initialisation performed for each instance? Individual codec `.init()` implementations might need to check whether an instance has already initialised on this card, but can we really skip this completely globally?

Reply:
@bardliao I guess this is ok as this init() is specific to the sof-sdw driver and you only do dailink specific actions, but it's worth a check still. I.e. is this future proof to not all init for all codecs on the dailink?

Reply:
> I don't understand the process well enough, but this raises an alarm for me. Are we really sure that it's always enough to just call `.init()` once for the first codec iinfo? That we'll never have mixed codecs or a codec, that needs certain initialisation performed for each instance? Individual codec `.init()` implementations might need to check whether an instance has already initialised on this card, but can we really skip this completely globally?

@lyakh If we look at the "codec init" functions, we can see that they are used to set dailink init function. So it is fine to called them once per dailink. We don't support mixed codecs in a dailink for now. If we want to do that, we will do more modification than this PR. Current code doesn't allow us to have different init function for the same codec either. So, I think the PR is safe for current code.

> @bardliao I guess this is ok as this init() is specific to the sof-sdw driver and you only do dailink specific actions, but it's worth a check still. I.e. is this future proof to not all init for all codecs on the dailink?

@kv2019i Yes, I checked all the init functions for supported codecs. None of them need to be called more than once per dailink. In fact, sof_sdw_rt1308_init() was the only one which will be called more than one time per dailink before this change.